File: java/Samples/BiometricStandards/VB/FCRecordSampleVB/MainForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports Neurotec.Biometrics.Standards
Imports Neurotec.Gui
Imports Neurotec.Images
Imports Neurotec.IO

Partial Public Class MainForm
	Inherits Form
#Region "Private static fields"

	Private Shared ansiBdbFormat As UInteger = BdifTypes.MakeFormat(CbeffBiometricOrganizations.IncitsTCM1Biometrics, CbeffBdbFormatIdentifiers.IncitsTCM1BiometricsFaceImage)
	Private Shared isoBdbFormat As UInteger = BdifTypes.MakeFormat(CbeffBiometricOrganizations.IsoIecJtc1SC37Biometrics, CbeffBdbFormatIdentifiers.IsoIecJtc1SC37BiometricsFaceImage)

#End Region

#Region "Private static methods"

	Private Shared Function GetOptions(ByVal mode As BdifOptionsForm.BdifOptionsFormMode, ByRef standard As BdifStandard, ByRef flags As UInteger, ByRef version As NVersion) As Boolean
		Using form As New FCRecordOptionsForm()
			form.Standard = standard
			form.Flags = flags
			form.Mode = mode
			form.Version = version
			If form.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				standard = form.Standard
				flags = form.Flags
				version = form.Version
			Else
				Return False
			End If
		End Using
		Return True
	End Function

	Private Shared Function GetOptions(ByRef format As UInteger) As Boolean
		Using form As New CbeffRecordOptionsForm()
			If form.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				format = form.PatronFormat
			Else
				Return False
			End If
		End Using
		Return True
	End Function

	Private Shared Function LoadTemplate(ByVal openFileDialog As OpenFileDialog, <System.Runtime.InteropServices.Out()> ByRef fileName As String) As Byte()
		openFileDialog.FileName = Nothing
		If openFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
			fileName = openFileDialog.FileName
			Return File.ReadAllBytes(openFileDialog.FileName)
		End If
		fileName = Nothing
		Return Nothing
	End Function

	Private Shared Sub SaveTemplate(ByVal saveFileDialog As SaveFileDialog, ByVal fileName As String, ByVal template() As Byte)
		saveFileDialog.FileName = fileName
		If saveFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
			File.WriteAllBytes(saveFileDialog.FileName, template)
		End If
	End Sub

	Private Shared Sub GetBiometricDataBlock(ByVal node As TreeNode, ByVal record As CbeffRecord)
		If record.BdbBuffer IsNot Nothing AndAlso (record.BdbFormat = ansiBdbFormat OrElse record.BdbFormat = isoBdbFormat) Then
			Dim standard As BdifStandard = If(record.BdbFormat = ansiBdbFormat, BdifStandard.Ansi, BdifStandard.Iso)
			Dim fcRecord As New FCRecord(record.BdbBuffer, standard)
			Dim child As New TreeNode("FCRecord")
			child.Tag = fcRecord
			AddFaceImages(child, fcRecord)
			node.Nodes.Add(child)
		End If
	End Sub

	Private Shared Sub AddFaceImages(ByVal node As TreeNode, ByVal record As CbeffRecord)
		GetBiometricDataBlock(node, record)

		For Each child As CbeffRecord In record.Records
			AddFaceImages(node, child)
		Next child
	End Sub

	Private Shared Function AddFaceImage(ByVal node As TreeNode, ByVal faceImage As FcrFaceImage) As TreeNode
		Dim index As Integer = node.Nodes.Count + 1
		Dim recordNode As New TreeNode("FaceImage " & index)
		recordNode.Tag = faceImage
		node.Nodes.Add(recordNode)
		Return recordNode
	End Function

	Private Shared Sub AddFaceImages(ByVal node As TreeNode, ByVal record As FCRecord)
		For Each faceImage As FcrFaceImage In record.FaceImages
			AddFaceImage(node, faceImage)
		Next faceImage
	End Sub

	Private Shared Function ConvertToStandard(ByVal record As FCRecord, ByVal newStandard As BdifStandard, ByVal flags As UInteger, ByVal version As NVersion) As FCRecord
		If record.Standard = newStandard AndAlso record.Flags = flags AndAlso record.Version = version Then
			Return record
		End If

		Return New FCRecord(record, flags, newStandard, version)
	End Function

#End Region

#Region "Private fields"

	Private _fileName As String

#End Region

#Region "Public constructor"

	Public Sub New()
		InitializeComponent()

		imageOpenFileDialog.Filter = NImages.GetOpenFileFilterString(True, True)
		rawImageOpenFileDialog.Filter = String.Format("All Supported Files ({0};{1})|{0};{1}|JPEG Files ({0})|{0}|JPEG 2000 Files ({1})|{1}|All Files (*.*)|*.*", NImageFormat.Jpeg.FileFilter, NImageFormat.Jpeg2K.FileFilter)
		saveImageFileDialog.Filter = NImages.GetSaveFileFilterString()
		aboutToolStripMenuItem.Text = "&"c + AboutBox.Name

		OnSelectedItemChanged()
	End Sub

#End Region

#Region "Private methods"

	Private Sub ShowError(ByVal message As String)
		MessageBox.Show(message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
	End Sub

	Private Function GetFirstNodeWithTag(Of T)() As TreeNode
		Dim node As TreeNode = treeView.SelectedNode
		If node Is Nothing Then
			Return Nothing
		End If

		Do While node.Parent IsNot Nothing AndAlso Not (TypeOf node.Tag Is T)
			node = node.Parent
		Loop

		Return If(TypeOf node.Tag Is T, node, Nothing)
	End Function

	Private Sub SetTemplate(ByVal record As FCRecord, ByVal fileName As String)
		Dim newRecord As Boolean = True
		Dim recordNode As TreeNode = GetFirstNodeWithTag(Of FCRecord)()
		If recordNode Is Nothing Then
			newRecord = True
		Else
			newRecord = (CType(recordNode.Tag, FCRecord)) IsNot record
		End If

		If newRecord Then
			_fileName = fileName
			treeView.BeginUpdate()
			treeView.Nodes.Clear()

			Dim templateNode As New TreeNode((If(fileName Is Nothing, "Untitled", Path.GetFileName(fileName))))
			templateNode.Tag = record

			AddFaceImages(templateNode, record)

			treeView.Nodes.Add(templateNode)
			treeView.SelectedNode = templateNode

			treeView.EndUpdate()
		End If
	End Sub

	Private Sub SetCbeff(ByVal record As CbeffRecord, ByVal fileName As String)
		Dim newRecord As Boolean = True
		Dim recordNode As TreeNode = GetFirstNodeWithTag(Of CbeffRecord)()
		If recordNode Is Nothing Then
			newRecord = True
		Else
			newRecord = (CType(recordNode.Tag, CbeffRecord)) IsNot record
		End If

		If newRecord Then
			_fileName = fileName
			treeView.BeginUpdate()
			treeView.Nodes.Clear()

			Dim root As New TreeNode((If(fileName Is Nothing, "Untitled", Path.GetFileName(fileName))))
			root.Tag = record

			AddFaceImages(root, record)

			treeView.Nodes.Add(root)
			treeView.SelectedNode = root

			treeView.EndUpdate()
		End If
	End Sub

	Private Sub NewTemplate()
		Dim standard As BdifStandard = BdifStandard.Iso
		Dim flags As UInteger = 0
		Dim version As NVersion = FCRecord.VersionIsoCurrent

		If (Not GetOptions(BdifOptionsForm.BdifOptionsFormMode.New, standard, flags, version)) Then
			Return
		End If

		NewTemplate(standard, flags, version)
	End Sub

	Private Sub NewTemplate(ByVal standard As BdifStandard, ByVal flags As UInteger, ByVal version As NVersion)
		Dim record As New FCRecord(standard, version, flags)
		SetTemplate(record, Nothing)
	End Sub

	Private Sub OpenTemplate()
		Dim fileName As String = Nothing
		Dim templ() As Byte = LoadTemplate(fcRecordOpenFileDialog, fileName)
		If templ Is Nothing Then
			Return
		End If

		Dim standard As BdifStandard = BdifStandard.Iso
		Dim flags As UInteger = 0
		Dim version As NVersion = FCRecord.VersionIsoCurrent
		If (Not GetOptions(BdifOptionsForm.BdifOptionsFormMode.Open, standard, flags, version)) Then
			Return
		End If

		Dim template As FCRecord
		Try
			template = New FCRecord(New NBuffer(templ), flags, standard)
		Catch ex As Exception
			ShowError("Failed to load template! Reason:" & Constants.vbCrLf & ex.ToString())
			Return
		End Try
		SetTemplate(template, fileName)
	End Sub

	Private Sub OpenCbeff()
		Dim fileName As String = Nothing
		Dim templ() As Byte = LoadTemplate(cbeffRecordOpenFileDialog, fileName)
		If templ Is Nothing Then
			Return
		End If

		Dim format As UInteger = 0
		If (Not GetOptions(format)) Then
			Return
		End If

		Dim template As CbeffRecord
		Try
			template = New CbeffRecord(New NBuffer(templ), format)
		Catch ex As Exception
			ShowError("Failed to load template! Reason:" & Constants.vbCrLf & ex.ToString())
			Return
		End Try

		SetCbeff(template, fileName)
	End Sub

	Private Sub SaveTemplate(ByVal record As FCRecord)
		SaveTemplate(fcRecordSaveFileDialog, _fileName, record.Save().ToArray())
		If TypeOf treeView.Nodes(0).Tag Is FCRecord Then
			treeView.Nodes(0).Text = Path.GetFileName(fcRecordSaveFileDialog.FileName)
		End If
	End Sub

	Private Sub OnSelectedItemChanged()
		Dim selected As TreeNode = treeView.SelectedNode

		If selected IsNot Nothing Then
			Dim faceImage As FcrFaceImage = TryCast(selected.Tag, FcrFaceImage)
			fcView.Record = faceImage

			If faceImage IsNot Nothing Then
				propertyGrid.SelectedObject = faceImage
			Else
				propertyGrid.SelectedObject = If(treeView.SelectedNode Is Nothing, Nothing, treeView.SelectedNode.Tag)
			End If

			If TypeOf selected.Tag Is CbeffRecord Then
				addFaceFromImageToolStripMenuItem.Enabled = False
				addFaceFromRawToolStripMenuItem.Enabled = False
				saveFaceAsDataToolStripMenuItem.Enabled = False
				saveFaceToolStripMenuItem.Enabled = False
				convertToolStripMenuItem.Enabled = False
				removeToolStripMenuItem.Enabled = False
				saveToolStripMenuItem.Enabled = False
			Else
				addFaceFromImageToolStripMenuItem.Enabled = True
				addFaceFromRawToolStripMenuItem.Enabled = True
				saveFaceAsDataToolStripMenuItem.Enabled = True
				saveFaceToolStripMenuItem.Enabled = True
				convertToolStripMenuItem.Enabled = True
				removeToolStripMenuItem.Enabled = True
				saveToolStripMenuItem.Enabled = True
			End If
		End If
	End Sub

#End Region

#Region "Private form methods"

	Private Sub saveFaceAsDataToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles saveFaceAsDataToolStripMenuItem.Click
		Dim selected As TreeNode = treeView.SelectedNode
		If (selected IsNot Nothing) AndAlso (selected.Parent IsNot Nothing) Then
			Dim img As FcrFaceImage = TryCast(selected.Tag, FcrFaceImage)
			If img IsNot Nothing Then
				saveRawFileDialog.Filter = If((img.ImageDataType = FcrImageDataType.Jpeg), String.Format("JPEG Files ({0})|{0}", NImageFormat.Jpeg.FileFilter), String.Format("JPEG 2000 Files ({0})|{0}", NImageFormat.Jpeg2K.FileFilter))

				If saveRawFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
					Try
						File.WriteAllBytes(saveRawFileDialog.FileName, img.ImageData.ToArray())
					Catch ex As Exception
						ShowError("Failed to save face to data." & Constants.vbCrLf & "Reason: " & ex.ToString())
						Return
					End Try
				End If
			End If
		End If
	End Sub

	Private Sub saveFaceToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles saveFaceToolStripMenuItem.Click
		Dim selected As TreeNode = treeView.SelectedNode
		If (selected IsNot Nothing) AndAlso (selected.Parent IsNot Nothing) Then
			If saveImageFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				Try
					Dim img As FcrFaceImage = TryCast(selected.Tag, FcrFaceImage)
					If img IsNot Nothing Then
						Using image As NImage = img.ToNImage()
							image.Save(saveImageFileDialog.FileName)
						End Using
					End If
				Catch ex As Exception
					ShowError("Failed to save image to file!" & Constants.vbCrLf & "Reason: " & ex.ToString())
					Return
				End Try
			End If
		End If
	End Sub

	Private Sub addFaceFromImageToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles addFaceFromImageToolStripMenuItem.Click
		Dim recordNode As TreeNode = GetFirstNodeWithTag(Of FCRecord)()
		If recordNode Is Nothing Then
			ShowError("Operation is not supported." & Constants.vbCrLf & "Add FCRecord first")
			Return
		End If

		If imageOpenFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
			Dim imageType As FcrFaceImageType
			Dim dataType As FcrImageDataType

			Using form As New AddFaceImageForm()
				Dim extension As String = Path.GetExtension(imageOpenFileDialog.FileName)
				dataType = If((NImageFormat.Jpeg2K.FileFilter).Contains(extension), FcrImageDataType.Jpeg2000, FcrImageDataType.Jpeg)
				form.ImageDataType = dataType
				If form.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
					imageType = form.FaceImageType
					dataType = form.ImageDataType
				Else
					Return
				End If
			End Using

			Try
				Using image As NImage = NImage.FromFile(imageOpenFileDialog.FileName)
					Dim record As FCRecord = TryCast(recordNode.Tag, FCRecord)

					Dim img As New FcrFaceImage(record.Standard, record.Version)
					img.FaceImageType = imageType
					img.ImageDataType = dataType
					img.SetImage(image)
					record.FaceImages.Add(img)
					Dim node As TreeNode = AddFaceImage(recordNode, img)
					treeView.SelectedNode = node
				End Using
			Catch ex As Exception
				ShowError("Failed to add image. Reason:" & Constants.vbCrLf & ex.ToString())
				Return
			End Try
			treeView.ExpandAll()
		End If
	End Sub

	Private Sub addFaceFromRawToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles addFaceFromRawToolStripMenuItem.Click
		If rawImageOpenFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
			Dim width, height As UShort
			Dim imageType As FcrFaceImageType
			Dim dataType As FcrImageDataType
			Dim imageColorSpace As FcrImageColorSpace
			Dim vendorImageColorSpace As Byte

			Using form As New RawFaceImageOptionsForm()
				Dim extension As String = Path.GetExtension(rawImageOpenFileDialog.FileName)
				dataType = If((NImageFormat.Jpeg2K.FileFilter).Contains(extension), FcrImageDataType.Jpeg2000, FcrImageDataType.Jpeg)
				form.ImageDataType = dataType

				If form.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
					imageType = form.FaceImageType
					dataType = form.ImageDataType
					imageColorSpace = form.ImageColorSpace
					width = form.ImageWidth
					height = form.ImageHeight
					vendorImageColorSpace = form.VendorColorSpace
				Else
					Return
				End If
			End Using

			Try
				Dim rawImage() As Byte = File.ReadAllBytes(rawImageOpenFileDialog.FileName)

				Dim recordNode As TreeNode = GetFirstNodeWithTag(Of FCRecord)()
				Dim record As FCRecord = TryCast(recordNode.Tag, FCRecord)

				Dim img As New FcrFaceImage(record.Standard, record.Version)
				img.FaceImageType = imageType
				img.ImageDataType = dataType
				img.Width = width
				img.Height = height
				img.SetImageColorSpace(imageColorSpace, vendorImageColorSpace)
				img.ImageData = New NBuffer(rawImage)
				record.FaceImages.Add(img)
				Dim node As TreeNode = AddFaceImage(treeView.Nodes(0), img)
				treeView.SelectedNode = node
			Catch ex As Exception
				ShowError("Failed to add image. Reason:" & Constants.vbCrLf & ex.ToString())
				Return
			End Try
		End If
	End Sub

	Private Sub removeToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles removeToolStripMenuItem.Click
		Dim selected As TreeNode = treeView.SelectedNode
		If selected Is Nothing Then
			Return
		End If

		Dim recordNode As TreeNode = GetFirstNodeWithTag(Of FCRecord)()
		If recordNode Is Nothing Then
			ShowError("Add FMRecord first")
			Return
		End If

		Dim fview As FcrFaceImage = TryCast(selected.Tag, FcrFaceImage)
		If fview IsNot Nothing Then
			Dim record As FCRecord = TryCast(recordNode.Tag, FCRecord)
			record.FaceImages.Remove(fview)
			recordNode.Nodes.Remove(selected)
		End If
	End Sub

	Private Sub convertToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles convertToolStripMenuItem.Click
		Dim rootNode As TreeNode = GetFirstNodeWithTag(Of FCRecord)()
		If rootNode Is Nothing Then
			Return
		End If

		Dim fcRecord As FCRecord = TryCast(rootNode.Tag, FCRecord)
		Dim standard As BdifStandard = If((fcRecord.Standard = BdifStandard.Ansi), BdifStandard.Iso, BdifStandard.Ansi)
		Dim flags As UInteger = 0
		Dim version As NVersion = If(standard = BdifStandard.Iso, fcRecord.VersionIsoCurrent, fcRecord.VersionAnsiCurrent)

		If (Not GetOptions(BdifOptionsForm.BdifOptionsFormMode.Convert, standard, flags, version)) Then
			Return
		End If

		fcRecord = ConvertToStandard(fcRecord, standard, flags, version)
		rootNode.Tag = fcRecord

		For i As Integer = 0 To fcRecord.FaceImages.Count - 1
			rootNode.Nodes(i).Tag = fcRecord.FaceImages(i)
		Next i

		OnSelectedItemChanged()
	End Sub

	Private Sub aboutToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles aboutToolStripMenuItem.Click
		AboutBox.Show()
	End Sub

	Private Sub newToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles newToolStripMenuItem.Click
		NewTemplate()
	End Sub

	Private Sub openToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles openToolStripMenuItem.Click
		OpenTemplate()
		treeView.ExpandAll()
	End Sub

	Private Sub openToolStripMenuItemCbeff_Click(ByVal sender As Object, ByVal e As EventArgs) Handles openToolStripMenuItemCbeff.Click
		OpenCbeff()
		treeView.ExpandAll()
	End Sub

	Private Sub saveToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles saveToolStripMenuItem.Click
		Dim recordNode As TreeNode = GetFirstNodeWithTag(Of FCRecord)()
		If recordNode Is Nothing Then
			ShowError("Add FCRecord first")
			Return
		End If

		Try
			SaveTemplate(CType(recordNode.Tag, FCRecord))
		Catch ex As Exception
			ShowError(ex.ToString())
		End Try
	End Sub

	Private Sub exitToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles exitToolStripMenuItem.Click
		Close()
	End Sub

	Private Sub treeView_AfterSelect(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles treeView.AfterSelect
		OnSelectedItemChanged()
	End Sub

#End Region
End Class

File: java/Samples/BiometricStandards/CS/FCRecordSampleCS/MainForm.cs
using System;
using System.IO;
using System.Windows.Forms;
using Neurotec.Biometrics.Common;
using Neurotec.Biometrics.Standards;
using Neurotec.Gui;
using Neurotec.Images;
using Neurotec.IO;

namespace Neurotec.Samples
{
	public partial class MainForm : Form
	{
		#region Private static fields

		private static uint ansiBdbFormat = BdifTypes.MakeFormat(CbeffBiometricOrganizations.IncitsTCM1Biometrics, CbeffBdbFormatIdentifiers.IncitsTCM1BiometricsFaceImage);
		private static uint isoBdbFormat = BdifTypes.MakeFormat(CbeffBiometricOrganizations.IsoIecJtc1SC37Biometrics, CbeffBdbFormatIdentifiers.IsoIecJtc1SC37BiometricsFaceImage);

		#endregion

		#region Private static methods

		private static bool GetOptions(BdifOptionsForm.BdifOptionsFormMode mode, ref BdifStandard standard, ref uint flags, ref NVersion version)
		{
			using (FCRecordOptionsForm form = new FCRecordOptionsForm())
			{
				form.Standard = standard;
				form.Flags = flags;
				form.Mode = mode;
				form.Version = version;
				if (form.ShowDialog() == DialogResult.OK)
				{
					standard = form.Standard;
					flags = form.Flags;
					version = form.Version;
				}
				else
				{
					return false;
				}
			}
			return true;
		}

		private static bool GetOptions(ref uint format)
		{
			using (CbeffRecordOptionsForm form = new CbeffRecordOptionsForm())
			{
				if (form.ShowDialog() == DialogResult.OK)
				{
					format = form.PatronFormat;
				}
				else
				{
					return false;
				}
			}
			return true;
		}

		private static byte[] LoadTemplate(OpenFileDialog openFileDialog, out string fileName)
		{
			openFileDialog.FileName = null;
			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				fileName = openFileDialog.FileName;
				return File.ReadAllBytes(openFileDialog.FileName);
			}
			fileName = null;
			return null;
		}

		private static void SaveTemplate(SaveFileDialog saveFileDialog, string fileName, byte[] template)
		{
			saveFileDialog.FileName = fileName;
			if (saveFileDialog.ShowDialog() == DialogResult.OK)
			{
				File.WriteAllBytes(saveFileDialog.FileName, template);
			}
		}

		private static void GetBiometricDataBlock(TreeNode node, CbeffRecord record)
		{
			if (record.BdbBuffer != null && (record.BdbFormat == ansiBdbFormat || record.BdbFormat == isoBdbFormat))
			{
				BdifStandard standard = record.BdbFormat == ansiBdbFormat ? BdifStandard.Ansi : BdifStandard.Iso;
				FCRecord fcRecord = new FCRecord(record.BdbBuffer, standard);
				TreeNode child = new TreeNode("FCRecord");
				child.Tag = fcRecord;
				AddFaceImages(child, fcRecord);
				node.Nodes.Add(child);
			}
		}

		private static void AddFaceImages(TreeNode node, CbeffRecord record)
		{
			GetBiometricDataBlock(node, record);

			foreach (CbeffRecord child in record.Records)
			{
				AddFaceImages(node, child);
			}
		}

		private static TreeNode AddFaceImage(TreeNode node, FcrFaceImage faceImage)
		{
			int index = node.Nodes.Count + 1;
			TreeNode recordNode = new TreeNode("FaceImage " + index);
			recordNode.Tag = faceImage;
			node.Nodes.Add(recordNode);
			return recordNode;
		}

		private static void AddFaceImages(TreeNode node, FCRecord record)
		{
			foreach (FcrFaceImage faceImage in record.FaceImages)
			{
				AddFaceImage(node, faceImage);
			}
		}

		private static FCRecord ConvertToStandard(FCRecord record, BdifStandard newStandard, uint flags, NVersion version)
		{
			if (record.Standard == newStandard && record.Flags == flags && record.Version == version)
				return record;

			return new FCRecord(record, flags, newStandard, version);
		}

		#endregion

		#region Private fields

		private string _fileName;

		#endregion

		#region Public constructor

		public MainForm()
		{
			InitializeComponent();

			imageOpenFileDialog.Filter = NImages.GetOpenFileFilterString(true, true);
			rawImageOpenFileDialog.Filter = string.Format("All Supported Files ({0};{1})|{0};{1}|JPEG Files ({0})|{0}|JPEG 2000 Files ({1})|{1}|All Files (*.*)|*.*", NImageFormat.Jpeg.FileFilter, NImageFormat.Jpeg2K.FileFilter);
			saveImageFileDialog.Filter = NImages.GetSaveFileFilterString();
			aboutToolStripMenuItem.Text = '&' + AboutBox.Name;

			OnSelectedItemChanged();
		}

		#endregion

		#region Private methods

		private void ShowError(string message)
		{
			MessageBox.Show(message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
		}

		private TreeNode GetFirstNodeWithTag<T>()
		{
			TreeNode node = treeView.SelectedNode;
			if (node == null)
				return null;

			while (node.Parent != null && !(node.Tag is T))
				node = node.Parent;

			return node.Tag is T ? node : null;
		}

		private void SetTemplate(FCRecord record, string fileName)
		{
			bool newRecord = true;
			TreeNode recordNode = GetFirstNodeWithTag<FCRecord>();
			if (recordNode == null)
				newRecord = true;
			else
				newRecord = ((FCRecord)recordNode.Tag) != record;

			if (newRecord)
			{
				_fileName = fileName;
				treeView.BeginUpdate();
				treeView.Nodes.Clear();

				TreeNode templateNode = new TreeNode((fileName == null ? "Untitled" : Path.GetFileName(fileName)));
				templateNode.Tag = record;

				AddFaceImages(templateNode, record);

				treeView.Nodes.Add(templateNode);
				treeView.SelectedNode = templateNode;

				treeView.EndUpdate();
			}
		}

		private void SetCbeff(CbeffRecord record, string fileName)
		{
			bool newRecord = true;
			TreeNode recordNode = GetFirstNodeWithTag<CbeffRecord>();
			if (recordNode == null)
				newRecord = true;
			else
				newRecord = ((CbeffRecord)recordNode.Tag) != record;

			if (newRecord)
			{
				_fileName = fileName;
				treeView.BeginUpdate();
				treeView.Nodes.Clear();

				TreeNode root = new TreeNode((fileName == null ? "Untitled" : Path.GetFileName(fileName)));
				root.Tag = record;

				AddFaceImages(root, record);

				treeView.Nodes.Add(root);
				treeView.SelectedNode = root;

				treeView.EndUpdate();
			}
		}

		private void NewTemplate()
		{
			BdifStandard standard = BdifStandard.Iso;
			uint flags = 0;
			NVersion version = FCRecord.VersionIsoCurrent;

			if (!GetOptions(BdifOptionsForm.BdifOptionsFormMode.New, ref standard, ref flags, ref version))
				return;

			NewTemplate(standard, flags, version);
		}

		private void NewTemplate(BdifStandard standard, uint flags, NVersion version)
		{
			FCRecord record = new FCRecord(standard, version, flags);
			SetTemplate(record, null);
		}

		private void OpenTemplate()
		{
			string fileName;
			byte[] templ = LoadTemplate(fcRecordOpenFileDialog, out fileName);
			if (templ == null)
				return;

			BdifStandard standard = BdifStandard.Iso;
			uint flags = 0;
			NVersion version = FCRecord.VersionIsoCurrent;
			if (!GetOptions(BdifOptionsForm.BdifOptionsFormMode.Open, ref standard, ref flags, ref version))
				return;

			FCRecord template;
			try
			{
				template = new FCRecord(new NBuffer(templ), flags, standard);
			}
			catch (Exception ex)
			{
				ShowError("Failed to load template! Reason:\r\n" + ex);
				return;
			}
			SetTemplate(template, fileName);
		}

		private void OpenCbeff()
		{
			string fileName;
			byte[] templ = LoadTemplate(cbeffRecordOpenFileDialog, out fileName);
			if (templ == null)
				return;

			uint format = 0;
			if (!GetOptions(ref format))
				return;

			CbeffRecord template;
			try
			{
				template = new CbeffRecord(new NBuffer(templ), format);
			}
			catch (Exception ex)
			{
				ShowError("Failed to load template! Reason:\r\n" + ex.ToString());
				return;
			}

			SetCbeff(template, fileName);
		}

		private void SaveTemplate(FCRecord record)
		{
			SaveTemplate(fcRecordSaveFileDialog, _fileName, record.Save().ToArray());
			if (treeView.Nodes[0].Tag is FCRecord)
				treeView.Nodes[0].Text = Path.GetFileName(fcRecordSaveFileDialog.FileName);
		}

		private void OnSelectedItemChanged()
		{
			TreeNode selected = treeView.SelectedNode;

			if (selected != null)
			{
				FcrFaceImage faceImage = selected.Tag as FcrFaceImage;
				fcView.Record = faceImage;

				if (faceImage != null)
				{
					propertyGrid.SelectedObject = faceImage;
				}
				else
				{
					propertyGrid.SelectedObject = treeView.SelectedNode == null ? null : treeView.SelectedNode.Tag;
				}

				if (selected.Tag is CbeffRecord)
				{
					addFaceFromImageToolStripMenuItem.Enabled = false;
					addFaceFromRawToolStripMenuItem.Enabled = false;
					saveFaceAsDataToolStripMenuItem.Enabled = false;
					saveFaceToolStripMenuItem.Enabled = false;
					convertToolStripMenuItem.Enabled = false;
					removeToolStripMenuItem.Enabled = false;
					saveToolStripMenuItem.Enabled = false;
				}
				else
				{
					addFaceFromImageToolStripMenuItem.Enabled = true;
					addFaceFromRawToolStripMenuItem.Enabled = true;
					saveFaceAsDataToolStripMenuItem.Enabled = true;
					saveFaceToolStripMenuItem.Enabled = true;
					convertToolStripMenuItem.Enabled = true;
					removeToolStripMenuItem.Enabled = true;
					saveToolStripMenuItem.Enabled = true;
				}
			}
		}

		#endregion

		#region Private form methods

		private void saveFaceAsDataToolStripMenuItem_Click(object sender, EventArgs e)
		{
			TreeNode selected = treeView.SelectedNode;
			if ((selected != null) && (selected.Parent != null))
			{
				FcrFaceImage img = selected.Tag as FcrFaceImage;
				if (img != null)
				{
					saveRawFileDialog.Filter = (img.ImageDataType == FcrImageDataType.Jpeg) ? string.Format("JPEG Files ({0})|{0}", NImageFormat.Jpeg.FileFilter) : string.Format("JPEG 2000 Files ({0})|{0}", NImageFormat.Jpeg2K.FileFilter);

					if (saveRawFileDialog.ShowDialog() == DialogResult.OK)
					{
						try
						{
							File.WriteAllBytes(saveRawFileDialog.FileName, img.ImageData.ToArray());
						}
						catch (Exception ex)
						{
							ShowError("Failed to save face to data.\r\nReason: " + ex);
							return;
						}
					}
				}
			}
		}

		private void saveFaceToolStripMenuItem_Click(object sender, EventArgs e)
		{
			TreeNode selected = treeView.SelectedNode;
			if ((selected != null) && (selected.Parent != null))
			{
				if (saveImageFileDialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						FcrFaceImage img = selected.Tag as FcrFaceImage;
						if (img != null)
							using (NImage image = img.ToNImage())
							{
								image.Save(saveImageFileDialog.FileName);
							}
					}
					catch (Exception ex)
					{
						ShowError("Failed to save image to file!\r\nReason: " + ex);
						return;
					}
				}
			}
		}

		private void addFaceFromImageToolStripMenuItem_Click(object sender, EventArgs e)
		{
			TreeNode recordNode = GetFirstNodeWithTag<FCRecord>();
			if (recordNode == null)
			{
				ShowError("Operation is not supported.\r\nAdd FCRecord first");
				return;
			}

			if (imageOpenFileDialog.ShowDialog() == DialogResult.OK)
			{
				FcrFaceImageType imageType;
				FcrImageDataType dataType;

				using (AddFaceImageForm form = new AddFaceImageForm())
				{
					string extension = Path.GetExtension(imageOpenFileDialog.FileName);
					dataType = (NImageFormat.Jpeg2K.FileFilter).Contains(extension) ? FcrImageDataType.Jpeg2000 : FcrImageDataType.Jpeg;
					form.ImageDataType = dataType;
					if (form.ShowDialog() == DialogResult.OK)
					{
						imageType = form.FaceImageType;
						dataType = form.ImageDataType;
					}
					else
					{
						return;
					}
				}

				try
				{
					using (NImage image = NImage.FromFile(imageOpenFileDialog.FileName))
					{
						FCRecord record = recordNode.Tag as FCRecord;

						FcrFaceImage img = new FcrFaceImage(record.Standard, record.Version);
						img.FaceImageType = imageType;
						img.ImageDataType = dataType;
						img.SetImage(image);
						record.FaceImages.Add(img);
						TreeNode node = AddFaceImage(recordNode, img);
						treeView.SelectedNode = node;
					}
				}
				catch (Exception ex)
				{
					ShowError("Failed to add image. Reason:\r\n" + ex);
					return;
				}
				treeView.ExpandAll();
			}
		}

		private void addFaceFromRawToolStripMenuItem_Click(object sender, EventArgs e)
		{
			if (rawImageOpenFileDialog.ShowDialog() == DialogResult.OK)
			{
				ushort width, height;
				FcrFaceImageType imageType;
				FcrImageDataType dataType;
				FcrImageColorSpace imageColorSpace;
				byte vendorImageColorSpace;

				using (RawFaceImageOptionsForm form = new RawFaceImageOptionsForm())
				{
					string extension = Path.GetExtension(rawImageOpenFileDialog.FileName);
					dataType = (NImageFormat.Jpeg2K.FileFilter).Contains(extension) ? FcrImageDataType.Jpeg2000 : FcrImageDataType.Jpeg;
					form.ImageDataType = dataType;

					if (form.ShowDialog() == DialogResult.OK)
					{
						imageType = form.FaceImageType;
						dataType = form.ImageDataType;
						imageColorSpace = form.ImageColorSpace;
						width = form.ImageWidth;
						height = form.ImageHeight;
						vendorImageColorSpace = form.VendorColorSpace;
					}
					else
					{
						return;
					}
				}

				try
				{
					byte[] rawImage = File.ReadAllBytes(rawImageOpenFileDialog.FileName);

					TreeNode recordNode = GetFirstNodeWithTag<FCRecord>();
					FCRecord record = recordNode.Tag as FCRecord;

					FcrFaceImage img = new FcrFaceImage(record.Standard, record.Version);
					img.FaceImageType = imageType;
					img.ImageDataType = dataType;
					img.Width = width;
					img.Height = height;
					img.SetImageColorSpace(imageColorSpace, vendorImageColorSpace);
					img.ImageData = new NBuffer(rawImage);
					record.FaceImages.Add(img);
					TreeNode node = AddFaceImage(treeView.Nodes[0], img);
					treeView.SelectedNode = node;
				}
				catch (Exception ex)
				{
					ShowError("Failed to add image. Reason:\r\n" + ex);
					return;
				}
			}
		}

		private void removeToolStripMenuItem_Click(object sender, EventArgs e)
		{
			TreeNode selected = treeView.SelectedNode;
			if (selected == null)
				return;

			TreeNode recordNode = GetFirstNodeWithTag<FCRecord>();
			if (recordNode == null)
			{
				ShowError("Add FMRecord first");
				return;
			}

			FcrFaceImage fview = selected.Tag as FcrFaceImage;
			if (fview != null)
			{
				FCRecord record = recordNode.Tag as FCRecord;
				record.FaceImages.Remove(fview);
				recordNode.Nodes.Remove(selected);
			}
		}

		private void convertToolStripMenuItem_Click(object sender, EventArgs e)
		{
			TreeNode rootNode = GetFirstNodeWithTag<FCRecord>();
			if (rootNode == null)
				return;

			FCRecord fcRecord = rootNode.Tag as FCRecord;
			BdifStandard standard = (fcRecord.Standard == BdifStandard.Ansi) ? BdifStandard.Iso : BdifStandard.Ansi;
			uint flags = 0;
			NVersion version = standard == BdifStandard.Iso ? FCRecord.VersionIsoCurrent : FCRecord.VersionAnsiCurrent;

			if (!GetOptions(BdifOptionsForm.BdifOptionsFormMode.Convert, ref standard, ref flags, ref version))
				return;

			fcRecord = ConvertToStandard(fcRecord, standard, flags, version);
			rootNode.Tag = fcRecord;

			for (int i = 0; i < fcRecord.FaceImages.Count; i++)
			{
				rootNode.Nodes[i].Tag = fcRecord.FaceImages[i];
			}

			OnSelectedItemChanged();
		}

		private void aboutToolStripMenuItem_Click(object sender, EventArgs e)
		{
			AboutBox.Show();
		}

		private void newToolStripMenuItem_Click(object sender, EventArgs e)
		{
			NewTemplate();
		}

		private void openToolStripMenuItem_Click(object sender, EventArgs e)
		{
			OpenTemplate();
			treeView.ExpandAll();
		}

		private void openToolStripMenuItemCbeff_Click(object sender, EventArgs e)
		{
			OpenCbeff();
			treeView.ExpandAll();
		}

		private void saveToolStripMenuItem_Click(object sender, EventArgs e)
		{
			TreeNode recordNode = GetFirstNodeWithTag<FCRecord>();
			if (recordNode == null)
			{
				ShowError("Add FCRecord first");
				return;
			}

			try
			{
				SaveTemplate((FCRecord)recordNode.Tag);
			}
			catch (Exception ex)
			{
				ShowError(ex.ToString());
			}
		}

		private void exitToolStripMenuItem_Click(object sender, EventArgs e)
		{
			Close();
		}

		private void treeView_AfterSelect(object sender, TreeViewEventArgs e)
		{
			OnSelectedItemChanged();
		}

		#endregion
	}
}
